astropy/io/fits/hdu/compressed.py

```diff
@@ -1605,6 +1605,9 @@ def header(self):
         for _ in range(table_blanks - image_blanks):
             image_header.append()
 
+        if 'SIMPLE' in image_header:
+            image_header.pop('XTENSION')
```

Comment:
Could you move this above, when XTENSION is written to the header ? I think we could just avoid writing this card with a `if 'ZSIMPLE' not in self._header`, line 1528.

Reply:
Sorry... I'm lost....
IIUC, `if 'ZSIMPLE' is not in self._header` the uncompressed image is not from Primary HDU.
The issue is when `'ZSIMPLE'` is in `self._header` and no `'ZTENSION'` is there, so that the compressed `'XTENSION'` is just kept (together with its value `'BINTABLE'`), not remapped in something else.
I moved the drop above as you asked, but it seems to me that I still have to check for 'SIMPLE' in the new header... am I wrong?

Reply:
Right, I was confused :D. But then that means that we could just delete XTENSION when we set SIMPLE right ? (So move the delete line in the `if 'ZSIMPLE' in self._header:` block)

Reply:
Right! who knows why I didn't do that in the first place! Knowledge is an iterative process apparently....

Reply:
I didn't think about it either ;)